### ISSUE TITLE
[2015 NIPS] Faster R-CNN: Towards Real-Time Object Detection with Region Proposal Networks

### ISSUE THREAD

**Issue by Jasonlee1995** [Jasonlee1995/AI_Papers#30]
![image](https://user-images.githubusercontent.com/49643709/124054937-f27ca200-da5d-11eb-9367-9c909a6707e9.png)

해당 논문은 Fast R-CNN에서 selective search를 통한 bounding box proposal을 neural network를 통해 학습하게 만듬

즉 RPN (Region Proposal Network)가 핵심이 됨

이에 RPN을 위주로 간단하게 설명

## 1. RPN (Region Proposal Network)
![image](https://user-images.githubusercontent.com/49643709/124055069-353e7a00-da5e-11eb-858f-1619b72474bf.png)

Fast R-CNN에서 region proposal로 사용되는 selective search 대신 학습할 수 있는 neural network를 통해 학습

Image가 feature extractor를 통해 생성된 feature map이 input임

해당 input으로부터 다양한 anchors를 가상으로 생성하여 fitting하여 feature map이 주어졌을 때 어떠한 candidate proposal을 생성하는게 좋은지 알려줌

### 1.1. RPN structure

해당 feature map이 n * n convolution과 2개의 1 * 1 convolution으로 구성 (regression, classification)

Fast R-CNN의 경우는 임의의 RoI에 대하여 fc layer를 통해 고정된 feature vector로 regression classification을 함

RPN의 경우는 window의 size가 일정하기에 해당되지 않으며, 1개의 regression이 1개의 anchor box regression을 책임짐

### 1.2. Anchor

Feature map이 n * n sliding window를 거칠 때, k개의 anchors를 생성하여 regression 진행

즉 k개의 anchors에 대하여 classification과 regression을 수행함 (anchor가 candidate인지 아닌지 + regression을 통해 anchor box를 수정)

참고로 학습할 때에는 image boundary를 넘어가는 anchor는 학습에서 제외

Test 시에는 제외하지 않고 진행하되, region proposal을 image boundary에 맞게 clipping

### 1.3. Loss function
![image](https://user-images.githubusercontent.com/49643709/124056124-1b9e3200-da60-11eb-8bfa-2578a8201f30.png)
![image](https://user-images.githubusercontent.com/49643709/124056160-2a84e480-da60-11eb-8d26-374c81ebb3c5.png)

요약하면 classification loss + regression loss로 구성

Regression loss의 경우 수가 더 많기에 (N_reg) lambda로 이를 조절 (lambda = 10 in paper)

### 1.4. Train

1개의 이미지로부터 256개의 anchors random sample함

비율은 최대한 positive와 negative의 비율이 1대 1이 되도록 맞춰줌 (모자르면 negative로 채움)

Positive anchor은 다음과 같이 정의
1. GT box와 가장 IoU가 높은 anchor인 경우
2. GT box와 IoU가 0.7 이상인 anchor의 경우

## 2. Inference

RPN을 통해 region proposal을 하면 대략 6,000개가 생성됨

이를 NMS를 통해 2,000개로 줄임 (IoU threshold 0.7)

Inference를 할 때, 여기서 top-N개를 선택하여 해당 N개에 대하여 Fast R-CNN을 돌림
